File: queryPineconeAndGPT.ts
import { Pinecone } from "@pinecone-database/pinecone";
import { OpenAIEmbeddings } from "langchain/embeddings/openai";
import OpenAI from "openai";

export default async (
  client: Pinecone,
  indexName: string,
  prompt: string
) => {
  console.log("Querying Pinecone vector store...");

  const index = client.Index(indexName);
  const queryEmbedding = await new OpenAIEmbeddings().embedQuery(prompt);

  let queryResponse = await index.query({
    topK: 5,
    vector: queryEmbedding,
    includeMetadata: true,
    includeValues: true,
  });

  console.log(`Found ${queryResponse.matches.length} matches...`);

  if (queryResponse.matches.length) {
    const concatenatedPageContent = queryResponse.matches
      .map((match) => match.metadata.pageContent)
      .join(" ");

    const openaiClient = new OpenAI();

    const augmentedPrompt =
      `Using the following information, answer the prompt at end: ${concatenatedPageContent}
      \n========================\n
      ${prompt}`;

    const completion = await openaiClient.chat.completions.create({
      model: "gpt-4o-mini",
      messages: [
        {
          role: "system",
          content: "You are a helpful assistant created to answer question based on RAG retrieval."
        },
        {
          role: "user",
          content: augmentedPrompt
        },
      ],
    });

    const dumbCompletion = await openaiClient.chat.completions.create({
      model: "gpt-4o-mini",
      messages: [
        {
          role: "system",
          content: "You are a helpful assistant created to answer question based on RAG retrieval."
        },
        {
          role: "user",
          content: prompt
        },
      ],
    });

    console.log(`\nAugmented answer: ${completion.choices[0].message.content}`);
    console.log("\n==================================\n");
    console.log(`\nDumb answer: ${dumbCompletion.choices[0].message.content}`);
  } else {
    console.log("No matches.");
  }
};

File: python/query_pinecone_and_GPT.py
from langchain_community.embeddings import OpenAIEmbeddings
import openai
from pinecone.grpc import PineconeGRPC as Pinecone


async def query_pinecone_and_gpt(client: Pinecone, index_name: str, prompt: str):
    print("Querying Pinecone vector store...")

    index = client.Index(index_name)
    query_embedding = await OpenAIEmbeddings().embed_query(prompt)

    query_response = await index.query(
        {
            "topK": 10,
            "vector": query_embedding,
            "includeMetadata": True,
            "includeValues": True,
        }
    )

    print(f"Found {len(query_response['matches'])} matches...")

    if query_response["matches"]:
        concatenated_page_content = " ".join(
            match["metadata"]["pageContent"] for match in query_response["matches"]
        )

        openai.api_key = "your_openai_api_key"

        augmented_prompt = (
            f"Using the following information, answer the prompt at end: {concatenated_page_content}"
            f"\n========================\n"
            f"{prompt}"
        )

        completion = openai.ChatCompletion.create(
            model="gpt-4o-mini",
            messages=[
                {
                    "role": "system",
                    "content": "You are a helpful assistant created to answer questions based on RAG retrieval.",
                },
                {"role": "user", "content": augmented_prompt},
            ],
        )

        print(f"\nAnswer: {completion.choices[0].message['content']}")
    else:
        print("No matches.")
